custom-routes-vis/src/main/java/com/google/sps/servlets/GetDirectionsServlet.java

```diff
@@ -0,0 +1,120 @@
+
+package com.google.sps.servlets;
+
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.HttpURLConnection;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/get-directions")
+public class GetDirectionsServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String origin = request.getParameter("origin");
+    String destination = request.getParameter("destination");
+    String endpoint = request.getParameter("endpoint");
+
+    String apiKey = request.getParameter("apiKey");
+    String directionsApiKey = System.getenv("DIRECTIONS_API_KEY");
+    String key = apiKey == "" ? apiKey : directionsApiKey;  // This works but it seems backwards to me.
+
+    URLConnection connection;
+
+    switch (endpoint) {
+      case "compute-routes":
+        connection = sendPostRequestToRoutesApi(origin, destination, false, key);
+        break;
+      case "compute-routes-alpha":
+        connection = sendPostRequestToRoutesApi(origin, destination, true, key);
+        break;
+      default:  // Default to use the Directions API.
+        connection = sendGetRequestToDirectionsApi(origin, destination, key);
+        break;
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(readResponse(connection));
+  }
+
+  /** 
+   * Sends request to the Directions API and returns the connection.
+   * @param origin Origin of the requested routes in "lat,lng" format.
+   * @param destination Destination of the requested routes in "lat,lng" format.
+   * @param key The API key.
+   */
+  public URLConnection sendGetRequestToDirectionsApi(String origin, String destination, String key) 
+      throws IOException, MalformedURLException {
+    String originDestination = "&origin=" + origin + "&destination=" + destination;
+    URL directionsUrl = new URL("https://maps.googleapis.com/maps/api/directions/json?"
+          + "mode=driving"
+          + "&alternatives=true" + originDestination + "&key=" + key);
+    URLConnection connection = directionsUrl.openConnection();
+    connection.setRequestProperty("Accept-Charset", "UTF-8");
+    return connection;
+  }
+
+  /** 
+   * Sends request to the Routes Preferred API and returns the connection.
+   * @param origin Origin of the requested routes in "lat,lng" format.
+   * @param destination Destination of the requested routes in "lat,lng" format.
+   * @param alpha Whether to send the request to the alpha version of the API.
+   * @param key The API key.
+   */
+  public URLConnection sendPostRequestToRoutesApi(String origin, String destination, boolean alpha,
```

Comment:
Similar to Haylee's comment above, you return a connection so maybe rename this connectToRoutesApi.